### ISSUE TITLE
DCS-BUG-6 - Spawn ground units as hidden on map does not work

### ISSUE THREAD

**Issue by shadowze** [FlightControl-Master/MOOSE#676]
https://forums.eagle.ru/showthread.php?t=206070

I set-up to use MOOSE spawn with some AA units
using 
ZoneTableHARD = { ZONE:New( "Zone1 HARD" ), ZONE:New( "Zone2 HARD" ), ZONE:New( "Zone3 HARD" ), ZONE:New( "Zone4 HARD" ) }
TemplateTableHARD = { "HA", "HB", "HC", "HD", "HE", "HF" }
Spawn_Vehicle_1HARD = SPAWN:New( "Spawn Vehicle HARD" )
  :InitLimit( 30, 500 )
  :InitRandomizeTemplate( TemplateTableHARD ) 
  :InitRandomizeZones( ZoneTableHARD )
  :SpawnScheduled( 40, .5 )

I have set-up the template vechs to be hidden on map
When they are spawned they are visible on the map

[Weapons_Training_v2.00.2.zip](https://github.com/FlightControl-Master/MOOSE/files/1275354/Weapons_Training_v2.00.2.zip)

**Comment by FlightControl-Master** [comment]
I am afraid that this is a DCS bug. Don't know if it is already reported but will check with ED.
So if your template had the hidden flag on, it should spawn as hidden and it does not.
I will double check the MOOSE code, but to my knowledge it does not touch the hidden flag.

**Comment by shadowze** [comment]
Thanks for getting back to me Sven

I used to use MIST respawnInZone and it does honour the Hidden on map flag on the units to be respawned 
I was trying to get away fro using MIST and use only MOOSE for my missions

**Comment by FlightControl-Master** [comment]
aha! That is good info! So there is something wrong in my code then! I'll report back to you!

**Comment by shadowze** [comment]
MIST respawnInZone , works differntly than your spawn
In MIST the objects are on the map at game start, you put a trigger on all that group all dead and then use respawninzone to pop them back 

See attached mission (go to Zugdidi and observe the Blue tank it will kill any spawned units

[RespawnInZOne.Weapons_Training_v1.41.7.zip](https://github.com/FlightControl-Master/MOOSE/files/1277122/RespawnInZOne.Weapons_Training_v1.41.7.zip)
Rename file extension to miz after u download it

**Comment by 132nd-etcher** [comment]
This may be unrelated, but I also have experienced discrepancies in the settings for dynamically spawned units: some of them were not controllable by humans (F10 "Set path"), or the option to set radar on/off wasn't available, etc.

**Comment by shadowze** [comment]
@132nd-etcher , see this bug report here -> https://github.com/FlightControl-Master/MOOSE/issues/664

**Comment by FlightControl-Master** [comment]
@shadowze Question: Using the mist function, does it create NEW groups or just replace the existing group? I remember I've been searching for this more than a year ago, but never found the solution how to do this. Can you maybe upload here or direct me to a demo mission using the mist function that does exactly what you request? One that ONLY demonstrates the function, so not a full blown mission pls.

**Comment by FlightControl-Master** [comment]
Ignore, i see you already did exactly that, and attached a mission above. Will take a look.

**Comment by FlightControl-Master** [comment]
Looking at your mission, it definitely is a still a DCS bug that is the cause.
This mist function does a re-spawn, which is different than spawn. re-spawn will actually re-use the existing group and re-ignite it. Because you put a trigger DEAD on the group, the group will be **re-spawned** when it is dead and thus, it will never be dead. It will always be alive.

The method coalition.addGroup() of DCS world does two things:
1. It can spawn new groups.
2. It can re-spawn existing groups.

But there is a big difference in how DCS world handles the visibility of these groups.

New groups will always be shown, there is no way to "hide" these groups. Even when the "hidden" flag is on at the template (and i don't touch this), the group will always be shown.

Existing and respawned groups will keep their existing properties (so they stay hidden).

Now, the good news is, MOOSE has a lot of tools to support this functionality too ...
But at the moment not exactly as mist works, although this can be made...

**The SPAWN class** 
Is to create new GROUP objects. 
Thus, SPAWN will instantiate new GROUP objects (with an indexed name using #nnn and the units #nnn-mm) in the simulation.
However! SPAWN does also have a method called ReSpawn!!! SPAWN has methods that work with the indexes of the spawned GROUP objects. SPAWN is actually like a list of GROUP objects, that are dynamically spawned. The first index of the list is 1.
What if I adapt SPAWN a bit, and allow to ReSpawn() a GROUP with index **0**? The 0 would express the GROUP object that is used as the template??? As such, you could use SPAWN to do all the fancy stuff available in SPAWN, like spawn at Vec2, Vec3 in Zone and all the other things, just using the initial template group that is ... re-spawned... Thus, the GROUP object that you use as the **Template** would not be _late activated_ and would just be a normal group at mission start.

SPAWN would not spawn anything, it would just re-use the template group to "respawn".
This is how this could look like in coding:

```lua
Spawn = SPAWN:New( "Template" ) --- Not late activated
```

... a bit later in your logic at an event handler or something ...

```lua
-- Respawn the Template Group.
-- The 0 expresses to ReSpawn the Template GROUP object.
Spawn:SpawnInZone( PolygonZone, true, 0 ) 
```

The following methods could be used to re-spawn the Template group object in a similar fashion:

```lua
function SPAWN:SpawnInZone( Zone, RandomizeGroup, SpawnIndex )
function SPAWN:SpawnFromVec2( Vec2, SpawnIndex )
function SPAWN:SpawnFromVec3( Vec3, SpawnIndex )
function SPAWN:SpawnFromUnit( HostUnit, SpawnIndex )
function SPAWN:SpawnFromStatic( HostStatic, SpawnIndex )
function SPAWN:ReSpawn( SpawnIndex )
```

I think it might also be possible to use the Init... methods:

```lua
function SPAWN:InitRandomizeZones( SpawnZoneTable )
function SPAWN:InitRandomizePosition( RandomizePosition, OuterRadius, InnerRadius )
```

**The GROUP class**
The second method is to use the Respawn() method at the GROUP class.
Yes, also that is possible.
But then I would need to extend the Respawn() functions on the GROUP class to also allow for respawning in zones, vec3, vec2 etc.


I have a couple of side remarks to all of this:

- First, why is the DCS method coalition.addGroup not working as expected. It should respect the hidden flag.

- Second, doing all this is like a "workaround". I want you to evaluate if this effort is urgent enough to let me start reworking SPAWN a bit (with the risk something goes wrong), impacting mission of other users and with the risk of damaging. It would not be the first time this happens. SPAWN is a core module. And all of this because it is required to have an object hidden when dynamically spawned?

- Third, In the back of my mind, actually all the ReSpawn() functionality in SPAWN is actually a candidate for refactoring into the GROUP class. Ideally, GROUP should have the methods to Respawn and let SPAWN manage the array of GROUP objects... GROUP should have the methods to spawn a new GROUP object from "itself". There will come a day when I will adapt MOOSE to this implementation, but not now.

So, @shadowze, taking these parameters, which is your best option to go for? What is your preference?

I would definitely delay this implementation to release 2.3, which you can start using in beta (on master), but not in release 2.2, because I don't want to endanger the stability of the current release.

**Comment by FlightControl-Master** [comment]
I am a bit confused.... What is not working?
I tried to resimulate the issue with a simple test mission, and it works perfectly...

This spawns a new group, hidden...
It is visible on the battle field (jump in the watch tower in external view).
It is not visible on the map.

https://github.com/FlightControl-Master/MOOSE_MISSIONS/tree/Release/SPA%20-%20Spawning/SPA-025%20-%20Ground%20Ops%20-%20Spawn%20Hidden

```lua
-- Name: SPA-025 - Ground Ops - Spawn Hidden
-- Author: FlightControl
-- Date Created: 06 Sep 2017
--
-- # Situation:
--
-- At Gudauta spawn a ground vehicle, hidden
-- 
-- # Test cases:
-- 
-- 1. Observe that the ground vehicle is spawned and his hidden.

-- Tests Gudauta
-- -------------
-- Spawn a gound vehicle...
Spawn_Vehicle_1 = SPAWN:New( "Spawn Vehicle 1" )
Spawn_Group_1 = Spawn_Vehicle_1:Spawn()
```

@shadowze 

Was this error due to a MP setup?

**Comment by FlightControl-Master** [comment]
I did something more ...

I have done a SPAWN mission using InitRandomizeTemplate.
And again it spawns okay... Jump into the watch tower and observe either a hawk or a patriot launcher to be launched.

https://github.com/FlightControl-Master/MOOSE_MISSIONS/tree/Release/SPA%20-%20Spawning/SPA-026%20-%20Ground%20Ops%20-%20Spawn%20RandomizeTemplate%20Hidden

```lua
-- Name: SPA-026 - Ground Ops - Spawn RandomizeTemplate Hidden
-- Author: FlightControl
-- Date Created: 06 Sep 2017
--
-- # Situation:
--
-- At Gudauta spawn a ground vehicle, hidden, based on a randomized template
-- 
-- # Test cases:
-- 
-- 1. Observe that a random ground vehicle is spawned and his hidden.



-- Tests Gudauta
-- -------------
-- Spawn a gound vehicle...

Templates = { "A", "B" }

Spawn_Vehicle_1 = SPAWN:New( "Vehicle" )
Spawn_Vehicle_1:InitRandomizeTemplate( Templates )
Spawn_Group_1 = Spawn_Vehicle_1:Spawn()
```

**Comment by shadowze** [comment]
Hmmm, I am using as stated randomisezone and randomisetemplates together,  did the AA spawn near Zugdidi hidden for you ? (on my supplied mission)
Because on the server they do not

**Comment by FlightControl-Master** [comment]
Can you take one of those missions I posted and run on your server?

**Comment by shadowze** [comment]
not right now people are playing (only a few mind you)
BUT I can tell you this
1. IF I open mission editor and load my mission run it for a bit check the log I can see HARD groups being spawned and they are NOT on the map
2. I host my mission on a "dedicated" server and I connect the AI HARD are spawning BUT are visible on the map
That is the difference, self host = works , host on a dedicated server = not working

**Comment by shadowze** [comment]
Ok I loaded up your mission on a dedicated server
.............................................. and
......................... the spawned objects are visible on the map

**Comment by FlightControl-Master** [comment]
So in SP this is working and in MP on the server not????

**Comment by shadowze** [comment]
Correct
SP loading mission from editor and running it = Objects are hidden

MP (dedicated server)  = Objects ignore the hidden on map check box and are visible in F10 view

**Comment by FlightControl-Master** [comment]
And are hidden units also visible on the client? The reason why asking is:
It would take time for me to test this. Maybe you know.
Preparing a report to Grimes.

**Comment by FlightControl-Master** [comment]
@shadowze sorry mate, but I can't simulate this!!!!

Run mission on a server, and went to spectators, and checked visibility **on the server** on the map, logged on as spectator. It does not show!!!

![afbeelding](https://user-images.githubusercontent.com/13690775/30151258-fd7f6182-93ae-11e7-88cb-958af39cedfa.png)

![afbeelding](https://user-images.githubusercontent.com/13690775/30151389-8abe1354-93af-11e7-84f5-3cdfbd54d116.png)

![afbeelding](https://user-images.githubusercontent.com/13690775/30151404-97a7cb14-93af-11e7-9344-64c1c3840318.png)

@shadowze 

Let us try to be absolutely clear... Are you watching on the SERVER machine, or connecting through a **CLIENT** to the SERVER???? I am spending 3 days on this already, and am losing time on this.

**Comment by FlightControl-Master** [comment]
@shadowze 

Please try to be more specific. This has been wasting my time due to an incomplete error report.

The **setup** needs to be explained well.

Let me explain you how to report this.

_I am connecting through a client machine (my PC) to a dedictated server. On the client machine I noticed that units that are dynamically spawned using MOOSE are not hidden on the map, although in the mission all the units are defined as hidden in the mission editor._

This would have helped a lot from the start :-)

Now, to resolve this issue, there is a problem. This is a **DCS BUG**. 
Dynamically spawned units **with the hidden flag on** show up on client machines, and should be hidden.

And I think I understand why the mist solution seems to work better. This is a deception.
Mist does not work better, but the process you apply in your mission is different.

In your mission, the AA unit is **NOT dynamically spawned**. It is just created normally as defined in the mission editor. When handling the dead event, what the function does is not spawn a new unit, but "teleport" the existing unit to a new place.

And it seems that in this case, the unit **stays hidden on the map**... Even when you reconnect, although I am not sure if this is the case.

Now, to develop a solution for this, please look a few posts above where I've written a comprehensive workaround solution to accomodate this kind of problem... Which by the way should not be written at all if DCS work work properly. I will do an error report to Grimes.

FC

**Comment by FlightControl-Master** [comment]
I've done a report to Grimes here:

https://forums.eagle.ru/showthread.php?p=3239306#post3239306

**Comment by shadowze** [comment]
Sorry for wasting your time Sven

I should be better at reporting issues
I have been testing software since the mid late 90's , its what I do for a living

**Comment by FlightControl-Master** [comment]
its okay. just try to give it a bit more attention. my time is more limited now, that's all.
thanks for reporting. You may help to show the urgency of it as Grimes will only see a report from me and won't see the urgency.

**Comment by FlightControl-Master** [comment]
Please consider my above suggestions as a workaround, and pls evaluate the urgency if it too. It can be done, and I don't think it is a whole lot of work, but it will require about 1 to 2 days of my attention to make this. Make = develop and test and document.

**Comment by FlightControl-Master** [comment]
Where would you prefer to have this development done, in GROUP or in SPAWN.

**Comment by shadowze** [comment]
Urgency , is not very urgent
As this is a DCS bug, we will wait and see if they fix it..........

I can use MIST respawninZone instead for now for this group and they will be hidden

**Comment by FlightControl-Master** [comment]
You will only see a fix appearing from ED if Grimes recognizes the problem and reports it, and chases it.
Right now, the bug report is non-existent to ED.

The only way to get it done is by raising the issue and get a community of people reacting to it.
Me alone reporting it won't help a lot i am afraid.

**Comment by FlightControl-Master** [comment]
@shadowze can you do me a favour, would you be able to make a small mission that spawns one new group, hidden, using mist? Then we can post it on ED forums and explain what is not working.

**Comment by FlightControl-Master** [comment]
never mind, Grimes already confirmed the error. It is reported to ED too. Hopefully they will go for a fix.

**Comment by shadowze** [comment]
closing out DCS Bug ... fuck knows when they will fix it

**Comment by FlightControl-Master** [comment]
We keep it....

**Comment by FlightControl-User** [comment]
This is a confirmed DCS bug. Cannot fix in release 2.3
